### ISSUE TITLE
Airlock 2FA on GrapheneOS?

### ISSUE THREAD

**Issue by jonathancross** [airlock/airlock-example-scripts#4]
Hello @orltom,
I was not able to find a repo for the actual Airlock 2FA Android app, so please excuse me adding this here.
I am trying to use this app on GrapheneOS (a fork of AOSP).
The app seems to be working fine with Google Play Services installed, but simply doesn't recognize the account setup QR code.
Any tips / suggestions would be appreciated.
Thanks!

Cross posted: https://www.reddit.com/r/GrapheneOS/comments/wpqhkm/airlock_2fa_app_on_grapheneos/

**Comment by mdoujak** [comment]
Hi jonathancross,

The Airlock 2FA app is officially supported only for devices running iOS or Android operating systems, and for the last, having the google services up and running. Nevertheless, we are aware of reports from a few users that were able to install and use the app in some alternative setups like for instance devices running the e OS ((https://e.foundation). We haven't heard from any user running GrapheneOS but it should be similar in most aspects. 

In your case, we have understood that the app seems to run correctly, but the enrollment seems to fail. Such failure can have multiple causes:

1. Time and date synchronization only allow a drift of a couple of seconds.
2. The QR Code to be scanned must be issued by a service that is integrated with the Airlock 2FA app. Some service providers use custom apps and the Airlock 2FA app will report an invalid QR code. 


Please note: the majority of the app's functionalities may work in such a setup, but some may not work. It may be that push notifications are not delivered since they rely on Google's Firebase messaging. You will still be able to authenticate by manually opening the app during authentication. 

I hope this helps. If not, can you provide us with more details as to what seems to be working and where the problems start?

mike

**Comment by jonathancross** [comment]
Thank you Mike!

> In your case, we have understood that the app seems to run correctly, but the enrollment seems to fail.

Well, it doesn't even fail... the QR code scanning for enrollment loads the camera, shows me the target, but simply doesn't recognize the QR code my bank is displaying.  No error is shown by the app at all, it just continues to act as though the QR code is not on screen yet. I've tried zooming (larger QR code), moving phone closer, farther holding still for 30+ seconds, etc.

The built-in QR code scanner works fine and scans the QR code in < 1 sec.

NOTE: This enrollment process works fine with Airlock 2FA running on an Android phone.

> Time and date synchronization only allow a drift of a couple of seconds.

Okay, that is a very tight window. ~~I'll ensure the clock is correct and try again.~~

Okay, I checked and time is perfect.

Still, I would expect an error message (eg "bad TOTP") if the time sync fails, correct?

> The QR Code to be scanned must be issued by a service that is integrated with the Airlock 2FA app. Some service providers use custom apps and the Airlock 2FA app will report an invalid QR code.

The bank is working with Airlock 2FA exclusively.  Enrollment and authentication are currently working perfectly from Android phones.  After a few minutes, the bank website shows an error and says to try again, but I think it is just a normal timeout.

Given that I can scan the QR code data (with the built-in app) -- Is it easy to decode this QR code value to something I can manually input into the app?
(I understand there might be security concerns, but I can ensure the seed is wiped from computer, etc)

> the majority of the app's functionalities may work in such a setup, but some may not work

Yes, of course. I'm prepared to deal with these small pieces of missing functionality.

Thanks!

**Edit: Updates made above after trying your suggestions.**

**Comment by jonathancross** [comment]
RE: Google's SafetyNet attestation service vs hardware attestation...
This might be helpful: https://grapheneos.org/articles/attestation-compatibility-guide

Generally, Android apps work perfectly on Graphene, unless they assume access to a feature which the OS restricts for privacy / security reasons.  Any suggestions where I might look?
Are these required for example?
- Internet
- Sensors
- Location
- IMEI
- Physical SIM
- Google account

**Comment by mdoujak** [comment]
Thanks for the additional information. 

> Generally, Android apps work perfectly on Graphene, unless they assume access to a feature which the OS restricts for privacy / security reasons. Any suggestions where I might look? Are these required for example?
> 
> * Sensors

I would suggest that you investigate the security settings for the sensors. It may be, that the Airlock 2FA is not able to scan the QR Code because it is not allowed to use the camera. This would of course work in the native camera app.

I hope this helps!
mike

**Comment by jonathancross** [comment]
The camera seems to be working fine, shows video in the app with overlay of the QR code "target", etc.
Does it use Google services somehow to recognize the QR code?

RE: Sensors...

No permissions are denied to the app.

I don't think there is anything I can necessarily do, but are there particular other sensors that the app uses?

> Given that I can scan the QR code data (with the built-in app) -- Is it easy to decode this QR code value to something I can manually input into the app?

Any info you can provide about decoding the QR data for manual input?

Thanks.

**Comment by mdoujak** [comment]
Our app uses the underlying features of the operating system to scan and interpret QR codes.

GrapheneOS provides this tiny bit of information on the camera:

> Google Camera
> Google Camera can be used with the [sandboxed Google Play compatibility layer](https://grapheneos.org/usage#sandboxed-google-play) and can take full advantage of the available cameras and image processing hardware as it can on the stock OS. It currently only depends on GSF and can be used without Play services (GMS) or the Play Store.

Since you wrote earlier that the Airlock 2FA app seems to work with the "Google Play Services" installed, I think you already have this compatibility layer. So as a result, I would expect the app to work, even if Graphene OS is not on our list of supported operating systems. 

Unfortunately, the implementation of an enrollment service that does not require a camera, is still quite some time in the future. 

I am sorry that I was not able to be of more assistance to resolve this issue. Should you gain more insights into what is causing your problem (e.g. local logs from your operating system) or even better find a working solution, please let us know.

regards, mike

**Comment by jonathancross** [comment]
> "Google Play Services" installed, I think you already have this compatibility layer. So as a result, I would expect the app to work...

Yes, it is installed and setup.  I also expected it to just work.

> Unfortunately, the implementation of an enrollment service that does not require a camera, is still quite some time in the future.

The app has a tab for "Manual Entry" of an "activation code"... Is this not a way to enroll without QR code?

**Comment by mdoujak** [comment]
Yes, the app already supports this functionality but the server does not (yet) support it. As I wrote, this feature is on the roadmap, but it is still quite some time in the future. I cannot be more specific since the implementation of the feature is one thing but then there is also deployment into the production environment of your specific service provider. This takes at least several months up to several years.

**Comment by LorenzBischof** [comment]
Any updates? I am experiencing the exact same behaviour. Google services are installed, the app and camera start without errors, but the QR code is not detected.

**Comment by datacyclist** [comment]
+1, exactly the same problem here, more than three years later.

I have a Pixel 8a with GrapheneOS installed, the bank is WIR bank and the airlock 2FA app fails to recognize the displayed QR code. The camera works just fine and shows the code that I'm supposed to scan. Another QR scanner recognizes the QR code as usual in a splitsecond. The airlock 2FA app has network, sensor and camera permission (and notifications, too). I'm on an unfiltered (at least on my side) 5G network connection.

I wished that the bank (other banks, too) had a regular 2FA that I could just put into Aegis where all my other 2FA accounts are, instead of trying to reinvent the 2FA again and again.

**Comment by datacyclist** [comment]
A workaround: airlock 2FA works on my GrapheneOS Pixel 8a if I switch to a different user that has the play services enabled. The app shows something about 'rooted device' but that's just the standard bs message. It works fine otherwise. I also get a 'Play Integrity API usage' notification -- which is probably what doesn't work in the standard user that doesn't use google play services (for millions of good reasons).

**Comment by ei8ht187** [comment]
Can confirm the findings of @datacyclist. Also a Bank WIR customer and have the bs "rooted device" message on my GrapheneOS Pixel 8 Pro with Play Service enabled. 
It would be great when you ditch the play integrity API and switch to the previously mentioned, Android hardware attestation API to fully and officially support GrapheneOS (and other ungoogled Android maybe too). Even better when you provide an APK download possibility outside the PlayStore. 
YUH already officially supports GrapheneOS with hardware attestation API: https://github.com/PrivSec-dev/banking-apps-compat-report/issues/509

**Comment by fabbra** [comment]
> A workaround: airlock 2FA works on my GrapheneOS Pixel 8a if I switch to a different user that has the play services enabled. The app shows something about 'rooted device' but that's just the standard bs message. It works fine otherwise. I also get a 'Play Integrity API usage' notification -- which is probably what doesn't work in the standard user that doesn't use google play services (for millions of good reasons).

I am running GrapheneOS on a Pixel 8a and having the same issues described above. 
However, I cannot make it run using the workaround mentioned. 

I am running Google PlayServices in the "main profile", still remain stuck with the QR code reader ~~and also do not get the "rooted device" message~~. 
I have installed the Airlock 2FA App via Aurora store.
Any ideas on how to fix this?
@datacyclist Did you install the App via the Google PlayStore? Would this make a difference?

@mdoujak is the "Activation Code" implemented on the server side in the meanwhile (3 years after your last message)?

Edit: After having properly initialized Google PlayServices (open Google PlayStore and tapping sign in but not logging in, [see here](https://kb.above.im/sandboxed-google-play-services/)) I also get the "rooted device" message but the QR code reader still is stuck and does not bring me any further.

**Comment by datacyclist** [comment]
@fabbra You mean where I got the airlock app from? That comes from the Aurora store, which is in any case just a privacy-enhanced mirror of the play store, so doesn't make a difference.

**Comment by mdoujak** [comment]
@fabbra: Yes, this functionality has been added quite a while ago already. We support enrollment of a device through an Activation Code that is provided to you either over the phone, on screen, or in a letter. 

It still depends on the configuration choices of your service provider (e.g,. your Bank), whether this option is enabled in their deployment.

**Comment by fabbra** [comment]
great, I will then try to get an activation code from my bank.

> [@fabbra](https://github.com/fabbra) You mean where I got the airlock app from? That comes from the Aurora store, which is in any case just a privacy-enhanced mirror of the play store, so doesn't make a difference.

for "the sake of completeness": @datacyclist in the google play store, are you signed in or using it "anonymously"?
I am currently NOT signed in to the google play store and - as describe above - the QR code reader does still not work (even though the "rooted device message" is now shown).

**Comment by datacyclist** [comment]
I was signed in with a google account in the separate profile. I just signed out now, started the app again, it still works. The message about the rooted device seems to be gone now 🤔 that's the opposite of what you're reporting.

**Comment by datacyclist** [comment]
> +1, exactly the same problem here, more than three years later.
> 
> I have a Pixel 8a with GrapheneOS installed, the bank is WIR bank and the airlock 2FA app fails to recognize the displayed QR code. ...

As a side note: Bank WIR has reset me to the SMS-based 2FA. And today I've been informed that they will go the route of having a mobile banking app which also acts as second factor for the web-based E-Banking. Airlock 2FA will be ditched completely. Let's hope the mobile banking app works on GrapheneOS.

**Comment by ei8ht187** [comment]
I already migrated to the new e-Banking. I can confirm that the new app runs flawless. There are even no false root warnings.
I also sent a note to the "DiBa" Team of Bank WIR, requesting support for the Android Hardware Attestation and a direct download possibilty without Play Store. You can do the same. Just reply to the message advertising the change to the new eBanking after you logged yourself in
.
Am Mittwoch, dem 26.11.2025 um 09:48 -0800 schrieb Georg Russ - notifications at github.com:

**Comment by exu-g** [comment]
@datacyclist Fyi, you can create a new 2FA device and still scan the code with Airlock. Looks like they just integrated the same solution into their own app.

The QR code only contains what looks like a secret, a UUID and the URL "api.futurae.com"
That seems to be the new app provider https://www.futurae.com/authenticators/

**Comment by datacyclist** [comment]
I've been notified that I'll be onboarded into their new E-Banking on Jan 13th (next week). Based on ei8ht187's comment, this should work flawlessly by then. Let's see. I'll send them some comments as suggested :-)

**Comment by datacyclist** [comment]
> I already migrated to the new e-Banking. I can confirm that the new app runs flawless. There are even no false root warnings.
> I also sent a note to the "DiBa" Team of Bank WIR, requesting support for the Android Hardware Attestation and a direct download possibilty without Play Store. You can do the same. Just reply to the message advertising the change to the new eBanking after you logged yourself in
> .
> Am Mittwoch, dem 26.11.2025 um 09:48 -0800 schrieb Georg Russ - notifications at github.com:
> […](#)

I tried to reply to the internal message, but it got rejected because the mailbox doesn't exist. Well, so I sent the message to the "Beratungsteam" instead. The new app and the E-Banking login work fine now on my Pixel 8a with GrapheneOS.
